treat Parameter the same way as Tensor

robieta opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #48966 Treat has_torch_function and object_has_torch_function as static False when scripting
* #48965 move has_torch_function to C++, and make a special case object_has_torch_function
* #48964 clean up imports for tensor.py
* **#48963 treat Parameter the same way as Tensor**

This PR makes the binding code treat `Parameter` the same way as `Tensor`, unlike all other `Tensor` subclasses. This does change the semantics of `THPVariable_CheckExact`, but it isn't used much and it seemed to make sense for the half dozen or so places that it is used.

Test plan: Existing unit tests. Benchmarks are in #48966

Differential Revision: [D25590733](https://our.internmc.facebook.com/intern/diff/D25590733)

Commit:
treat Parameter the same way as Tensor, and add fast path check in codegen

[ghstack-poisoned]

Commit:
Update on "treat Parameter the same way as Tensor, and add fast path check in codegen"


This PR makes the binding code treat `Parameter` the same way as `Tensor`, unlike all other `Tensor` subclasses. It also fixes an omission wherein Tensor methods that take no arguments were going through a suboptimal path checking for torch function.

This does change the semantics of `THPVariable_CheckExact`, but it isn't used much and it seemed to make sense for the half dozen or so places that it is used.

[ghstack-poisoned]

Commit:
Update on "treat Parameter the same way as Tensor, and add fast path check in codegen"


This PR makes the binding code treat `Parameter` the same way as `Tensor`, unlike all other `Tensor` subclasses. It also fixes an omission wherein Tensor methods that take no arguments were going through a suboptimal path checking for torch function.

This does change the semantics of `THPVariable_CheckExact`, but it isn't used much and it seemed to make sense for the half dozen or so places that it is used.

[ghstack-poisoned]

Commit:
Update on "treat Parameter the same way as Tensor, and add fast path check in codegen"


This PR makes the binding code treat `Parameter` the same way as `Tensor`, unlike all other `Tensor` subclasses. It also fixes an omission wherein Tensor methods that take no arguments were going through a suboptimal path checking for torch function.

This does change the semantics of `THPVariable_CheckExact`, but it isn't used much and it seemed to make sense for the half dozen or so places that it is used.

[ghstack-poisoned]

albanD commented on tools/autograd/gen_python_functions.py:
```diff
@@ -515,7 +515,7 @@ def gen_has_torch_function_check(
         if method:
             return f"""\
-if(check_has_torch_function(self_)) {{
+if(!THPVariable_CheckExact(self_) && check_has_torch_function(self_)) {{
```
Why do you need this extra check here?

robieta replied:
This is a fast path. I've come to the conclusion that because I've written something very similar to `check_has_torch_function` in https://github.com/pytorch/pytorch/pull/48965, the logic really should be merged. Which probably means the check will be removed from this PR, and `check_torch_function` will be replaced later in the PR stack.

Commit:
Update on "treat Parameter the same way as Tensor, and add fast path check in codegen"


This PR makes the binding code treat `Parameter` the same way as `Tensor`, unlike all other `Tensor` subclasses. It also fixes an omission wherein Tensor methods that take no arguments were going through a suboptimal path checking for torch function.

This does change the semantics of `THPVariable_CheckExact`, but it isn't used much and it seemed to make sense for the half dozen or so places that it is used.

[ghstack-poisoned]

Commit:
Update on "treat Parameter the same way as Tensor, and add fast path check in codegen"


This PR makes the binding code treat `Parameter` the same way as `Tensor`, unlike all other `Tensor` subclasses. It also fixes an omission wherein Tensor methods that take no arguments were going through a suboptimal path checking for torch function.

This does change the semantics of `THPVariable_CheckExact`, but it isn't used much and it seemed to make sense for the half dozen or so places that it is used.

[ghstack-poisoned]

Commit:
Update on "treat Parameter the same way as Tensor"


This PR makes the binding code treat `Parameter` the same way as `Tensor`, unlike all other `Tensor` subclasses. This does change the semantics of `THPVariable_CheckExact`, but it isn't used much and it seemed to make sense for the half dozen or so places that it is used.

[ghstack-poisoned]

robieta commented:
Ok, this is ready for review.

hameerabbasi commented on torch/csrc/autograd/init.cpp:
```diff
@@ -35,6 +35,14 @@ PyObject* THPAutograd_initExtension(PyObject* _unused, PyObject *unused) {
   auto _C_m = py::handle(torch_C_module).cast<py::module>();
   auto m = _C_m.def_submodule("_autograd", "autograd bindings");
 
+  auto parameter_module = THPObjectPtr(PyImport_ImportModule("torch.nn.parameter"));
```
Doesn't really matter much, but out of curiosity, is this a `py::object`?

robieta replied:
It looks like there's an analog from TH. (Based on some grep-ing it seems to just be handling the incref and decref.) I just cargo cult'd the other ones.

hameerabbasi reviewed (approved):
LGTM!

Commit:
Update on "treat Parameter the same way as Tensor"


This PR makes the binding code treat `Parameter` the same way as `Tensor`, unlike all other `Tensor` subclasses. This does change the semantics of `THPVariable_CheckExact`, but it isn't used much and it seemed to make sense for the half dozen or so places that it is used.

Test plan: Existing unit tests. Benchmarks are in #48966

[ghstack-poisoned]

Commit:
Update on "treat Parameter the same way as Tensor"


This PR makes the binding code treat `Parameter` the same way as `Tensor`, unlike all other `Tensor` subclasses. This does change the semantics of `THPVariable_CheckExact`, but it isn't used much and it seemed to make sense for the half dozen or so places that it is used.

Test plan: Existing unit tests. Benchmarks are in #48966

Differential Revision: [D25590733](https://our.internmc.facebook.com/intern/diff/D25590733)

[ghstack-poisoned]

Commit:
Update on "treat Parameter the same way as Tensor"


This PR makes the binding code treat `Parameter` the same way as `Tensor`, unlike all other `Tensor` subclasses. This does change the semantics of `THPVariable_CheckExact`, but it isn't used much and it seemed to make sense for the half dozen or so places that it is used.

Test plan: Existing unit tests. Benchmarks are in #48966

Differential Revision: [D25590733](https://our.internmc.facebook.com/intern/diff/D25590733)

[ghstack-poisoned]

Commit:
Update on "treat Parameter the same way as Tensor"


This PR makes the binding code treat `Parameter` the same way as `Tensor`, unlike all other `Tensor` subclasses. This does change the semantics of `THPVariable_CheckExact`, but it isn't used much and it seemed to make sense for the half dozen or so places that it is used.

Test plan: Existing unit tests. Benchmarks are in #48966

Differential Revision: [D25590733](https://our.internmc.facebook.com/intern/diff/D25590733)

[ghstack-poisoned]

Commit:
Update on "treat Parameter the same way as Tensor"


This PR makes the binding code treat `Parameter` the same way as `Tensor`, unlike all other `Tensor` subclasses. This does change the semantics of `THPVariable_CheckExact`, but it isn't used much and it seemed to make sense for the half dozen or so places that it is used.

Test plan: Existing unit tests. Benchmarks are in #48966

Differential Revision: [D25590733](https://our.internmc.facebook.com/intern/diff/D25590733)

[ghstack-poisoned]

ezyang commented:
It seems to me that more comments need to updated. For example:

```
bool is_tensor_and_append_overloaded(PyObject* obj, std::vector<py::handle>* overloaded_args) {
  if (THPVariable_CheckExact(obj)) {
    // torch.Tensor instances (not subclasses)
    return true;
  }
```

and probably a note on `THPVariable_CheckExact` saying what the intended use of this function is

ezyang reviewed (approved):
Going ahead and preemptively approving, but I do care about the comments quite a bit

Commit:
Update on "treat Parameter the same way as Tensor"


This PR makes the binding code treat `Parameter` the same way as `Tensor`, unlike all other `Tensor` subclasses. This does change the semantics of `THPVariable_CheckExact`, but it isn't used much and it seemed to make sense for the half dozen or so places that it is used.

Test plan: Existing unit tests. Benchmarks are in #48966

Differential Revision: [D25590733](https://our.internmc.facebook.com/intern/diff/D25590733)

[ghstack-poisoned]

Commit:
Update on "treat Parameter the same way as Tensor"


This PR makes the binding code treat `Parameter` the same way as `Tensor`, unlike all other `Tensor` subclasses. This does change the semantics of `THPVariable_CheckExact`, but it isn't used much and it seemed to make sense for the half dozen or so places that it is used.

Test plan: Existing unit tests. Benchmarks are in #48966

Differential Revision: [D25590733](https://our.internmc.facebook.com/intern/diff/D25590733)

[ghstack-poisoned]

Commit:
Update on "treat Parameter the same way as Tensor"


This PR makes the binding code treat `Parameter` the same way as `Tensor`, unlike all other `Tensor` subclasses. This does change the semantics of `THPVariable_CheckExact`, but it isn't used much and it seemed to make sense for the half dozen or so places that it is used.

Test plan: Existing unit tests. Benchmarks are in #48966

Differential Revision: [D25590733](https://our.internmc.facebook.com/intern/diff/D25590733)

[ghstack-poisoned]
